File: dotnet/System/workspace/csharp/allors.workspace.domain/data/Select.cs
// <copyright file="Select.cs" company="Allors bvba">
// Copyright (c) Allors bvba. All rights reserved.
// Licensed under the LGPL license. See LICENSE file in the project root for full license information.
// </copyright>

namespace Allors.Workspace.Data
{
    using System.Collections.Generic;
    using Meta;

    public class Select : IVisitable
    {
        public Select()
        {
        }

        public Select(params IPropertyType[] propertyTypes)
        {
            if (propertyTypes.Length > 0)
            {
                this.Step = new Step(propertyTypes, 0);
            }
        }

        public IEnumerable<Node> Include { get; set; }

        public Step Step { get; set; }

        public void Accept(IVisitor visitor) => visitor.VisitSelect(this);
    }
}

File: typescript/modules/libs/system/workspace/domain/src/lib/data/select.ts
import { Composite, PropertyType } from '@allors/system/workspace/meta';
import { Node } from '../pointer/node';

export interface Select {
  propertyType: PropertyType;

  ofType?: Composite;

  next?: Select;

  include?: Node[];
}

export function selectLeaf(select: Select): Select {
  if (select.next) {
    return selectLeaf(select.next);
  }

  return select;
}
